File: jwks-server.js
const express = require('express');
const jwt = require('jsonwebtoken');
const fs = require('fs');
const basicAuth = require('express-basic-auth');
const { generateKeyPair } = require('./keygen');
const cors = require('cors');
require('dotenv').config();
const port = process.env.PORT;
const jwt_key = process.env.JWT_KEY;
const app = express();
//const port = 8080;

app.use(cors());
app.use(express.static('public'));
//Middleware to parse JSON requests
app.use(express.json());

//Load existing keys from keys.json
const keys = JSON.parse(fs.readFileSync('keys.json', 'utf8'));

//JWKS endpoint
app.get('/jwks', (req, res) => {
    const currentTime = Date.now();
    const validKeys = keys.filter((key) => key.expiry > currentTime);
    const jwks = {
        keys: validKeys.map((key) => ({
            kid: key.kid,
            kty: 'RSA',
            nbf: key.expiry - 24 * 60 * 60 * 1000, //Not Before (1 day before expiry)
            use: 'sig',
            alg: 'RS256',
            e: 'AQAB',
            n: key.publicKey,
        })),
    };
    res.json(jwks);
        
});

//
app.get('/', (req, res) => {
  //health check route
  res.status(200).send({ code: 0, message: 'ok' });
});

app.get('/token', (req, res) =>{
  //route to get a token
  let id = Math.random().toString(36).substring(2, 8);
  let limit = 60 * 3; //180 secs
  let expires = Math.floor(Date.now() / 1000) + limit;
  let payload = {
    _id: id,
    exp: expires,
  };
  let token = jwt.sign(payload, jwt_key);
  res.status(201).send({ code:0, message: 'ok', data: token });
});


app.get('/test', (req, res) => {
  //simulate route that needs a valid token to access
  const header = req.header('Authorization');
  const [type, token] = header.split(' ');
  if (type === 'Bearer' && typeof token !== 'undefined') {
    try {
      let payload = jwt.verify(token, jwt_key);
      let current = Math.floor(Date.now() / 1000);
      let diff = current - payload.exp;
      res.status(200).send({ code: 0, message: `all good. ${diff} remaining` });
    } catch (err) {
      res.status(401).send({ code: 123, message: 'Invalid or expired token.' });
    }
  } else {
    res.status(401).send({ code: 456, message: 'Invalid token' });
  }
});
//



// Authentication endpoint
app.post('/auth', basicAuth({ users: { 'userABC': 'password123' } }), (req, res) => {
  const expired  = req.query;
  const keyToUse = expired ? keys[1] : keys[0]; // Use the second key if 'expired' is present

  const payload = {
    sub: 'userABC',
    // Add other claims as needed
  };

  const token = jwt.sign(payload, keyToUse.rsaPrivateKey, { algorithm: 'RS256', keyid: keyToUse.kid });

  res.json({ token });
});


app.listen(port, () => {
  console.log(`JWKS server is running on port ${port}`);
});

File: jwks-server.go
package main

import (
	"crypto/rand"
	"crypto/rsa"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"time"

	"github.com/dgrijalva/jwt-go"
	"github.com/gorilla/mux"
)

var rsaKey *rsa.PrivateKey

func init() {
	var err error
	rsaKey, err = rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		panic(err)
	}
}

func encodeToBase64URL(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

func bigIntToBytes(i *big.Int) []byte {
	return i.Bytes()
}

func jwksHandler(w http.ResponseWriter, r *http.Request) {
	keys := []JWK{}

	// Check if the RSA key is not nil
	if rsaKey != nil {
		key := JWK{
			Kty: "RSA",
			Kid: "your-key-identifier",
			Use: "sig",
			Alg: "RS256", // Replace with the appropriate algorithm
			N:   encodeToBase64URL(rsaKey.N.Bytes()),
			E:   encodeToBase64URL(bigIntToBytes(big.NewInt(int64(rsaKey.E)))),
		}
		keys = append(keys, key)
	}

	jwksResponse := JWKSResponse{
		Keys: keys,
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)

	if err := json.NewEncoder(w).Encode(jwksResponse); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}
type JWK struct {
	Kty string `json:"kty"`
	Kid string `json:"kid"`
	Use string `json:"use"`
	Alg string `json:"alg"`
	N   string `json:"n"`
	E   string `json:"e"`
}

type JWKSResponse struct {
	Keys []JWK `json:"keys"`
}

func main() {
	r := mux.NewRouter()
	// Define your routes
	r.HandleFunc("/jwks", jwksHandler).Methods("GET")
	r.HandleFunc("/auth", authHandler).Methods("POST")

	// Register your routes with the router
	http.Handle("/", r)

	fmt.Println("Server is running on :8080...")
	http.ListenAndServe(":8080", nil)
}

func authHandler(w http.ResponseWriter, r *http.Request) {
	expired := r.URL.Query().Get("expired")

	var token *jwt.Token
	if expired != "" {
		// If the "expired" query parameter is present, issue a JWT with an expired key
		token = generateKey("expired-key", time.Now().Add(-1*time.Hour))
	} else {
		// Issue a JWT with the current valid key
		token = generateKey("valid-key", time.Now().Add(24*time.Hour))
	}

	tokenString, err := token.SignedString(rsaKey)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/jwt")
	w.Write([]byte(tokenString))
}

func generateKey(kid string, expiry time.Time) *jwt.Token {
	claims := jwt.MapClaims{
		"kid": kid,
		"exp": expiry.Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	return token
}

